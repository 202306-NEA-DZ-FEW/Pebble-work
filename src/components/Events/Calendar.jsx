import Image from "next/image";
import Link from "next/link";
import React, { useState } from "react";

import styles from "@/styles/Calender.module.css";

const Calendar = () => {
    const [currentDate, setCurrentDate] = useState(new Date());

    const handleNextMonth = () => {
        setCurrentDate((prevDate) => {
            const nextMonth = new Date(
                prevDate.getFullYear(),
                prevDate.getMonth() + 1
            );
            return nextMonth;
        });
    };

    const handlePreviousMonth = () => {
        setCurrentDate((prevDate) => {
            const previousMonth = new Date(
                prevDate.getFullYear(),
                prevDate.getMonth() - 1
            );
            return previousMonth;
        });
    };

    const getMonthName = (date) => {
        const options = { month: "long" };
        return date.toLocaleDateString("en-US", options);
    };

    const renderCalendar = () => {
        const currentMonth = currentDate.getMonth();
        const currentYear = currentDate.getFullYear();

        const firstDay = new Date(currentYear, currentMonth, 1).getDay();
        const daysInMonth = new Date(
            currentYear,
            currentMonth + 1,
            0
        ).getDate();

        const calendarDays = [];
        const weekdays = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];

        // Render weekdays
        for (let i = 0; i < 7; i++) {
            calendarDays.push(
                <div
                    key={`weekday-${i}`}
                    className={`${styles.weekday} md:text-[12px] xl:text-[17px]`}
                >
                    {weekdays[i]}
                </div>
            );
        }

        // Render empty days before the first day of the month
        for (let i = 0; i < firstDay; i++) {
            calendarDays.push(
                <div key={`empty-${i}`} className={`${styles.emptyDay}`}></div>
            );
        }

        // Render day numbers
        for (let day = 1; day <= daysInMonth; day++) {
            calendarDays.push(
                <Link
                    href='#'
                    key={`day-${day}`}
                    className={`${styles.calendarDay} xl:text-[17px] md:w-[10px] xl:w-[18px] text-center md:text-[11px] rounded`}
                >
                    {day}
                </Link>
            );
        }

        return calendarDays;
    };

    return (
        <>
            <div className='flex flex-col-reverse items-center justify-center xl:w-[333px] xl:h-[243px] md:w-[222px] md:h-[180px]'>
                <div className='flex justify-center items-center gap-2'>
                    <button
                        className='flex items-center'
                        onClick={handlePreviousMonth}
                    >
                        <Image
                            className='xl:w-[50px] xl:h-[50px] md:w-[35px] md:h-[35px]'
                            src='/icons/previous.png'
                            width={50}
                            height={50}
                            alt='Previous'
                        />
                    </button>
                    <div className='flex flex-col items-center gap-3'>
                        <div className='calendar grid grid-cols-7 gap-1'>
                            {renderCalendar()}
                        </div>
                        <h2 className='text-xl font-bold md:text-[14px] xl:text-[18px] '>
                            {getMonthName(currentDate)}{" "}
                            {currentDate.getFullYear()}
                        </h2>
                    </div>

                    <button
                        className='flex items-center'
                        onClick={handleNextMonth}
                    >
                        <Image
                            className='xl:w-[50px] xl:h-[50px] md:w-[35px] md:h-[35px]'
                            src='/icons/next.png'
                            width={50}
                            height={50}
                            alt='Next'
                        />
                    </button>
                </div>
            </div>
        </>
    );
};

export default Calendar;
