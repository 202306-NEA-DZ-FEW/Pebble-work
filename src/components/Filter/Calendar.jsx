import Image from "next/image";

import React, { useState, useEffect } from "react";

import styles from "@/styles/Events.module.css";
import Dropdown from "./Dropdown";

// Define the Calendar component
const Calendar = ({ checkEvents, resetDays }) => {
    // State for the current date and selected days
    const [currentDate, setCurrentDate] = useState(new Date());
    const [selectedDays, setSelectedDays] = useState({});
    const [rangeStart, setRangeStart] = useState(null);
    const [selectedRange, setSelectedRange] = useState([]);

    // Function to handle date click
    const handleDateClick = (day) => {
        const selectedDate = new Date(
            currentDate.getFullYear(),
            currentDate.getMonth(),
            day + 1
        );

        if (rangeStart === null) {
            setRangeStart(selectedDate);
        } else {
            let startDate = rangeStart;
            let endDate = selectedDate;

            if (startDate > endDate) {
                [startDate, endDate] = [endDate, startDate];
            }

            const range = [];
            let currentDate = startDate;

            while (currentDate <= endDate) {
                range.push(new Date(currentDate));
                currentDate.setDate(currentDate.getDate() + 1);
            }

            setSelectedRange(range);
            console.log(range);

            checkEvents(range.map((date) => date.toISOString().split("T")[0]));
            setRangeStart(null);
        }
    };

    // handle next month button click
    const handleNextMonth = () => {
        setCurrentDate((prevDate) => {
            const nextMonth = new Date(
                prevDate.getFullYear(),
                prevDate.getMonth() + 1
            );
            return nextMonth;
        });
    };

    // handle previous month button click
    const handlePreviousMonth = () => {
        setCurrentDate((prevDate) => {
            const previousMonth = new Date(
                prevDate.getFullYear(),
                prevDate.getMonth() - 1
            );
            return previousMonth;
        });
    };

    // get the name of the month
    const getMonthName = (date) => {
        const options = { month: "long" };
        return date.toLocaleDateString("en-US", options);
    };

    // render the calendar
    const renderCalendar = () => {
        const currentMonth = currentDate.getMonth();
        const currentYear = currentDate.getFullYear();

        const firstDay = new Date(currentYear, currentMonth, 1).getDay();
        const daysInMonth = new Date(
            currentYear,
            currentMonth + 1,
            0
        ).getDate();

        const calendarDays = [];
        const weekdays = ["S", "M", "T", "W", "T", "F", "S"];

        // Render weekdays
        for (let i = 0; i < 7; i++) {
            calendarDays.push(
                <div
                    key={`weekday-${i}`}
                    className={`sm:text-[12px] font-[500] w-[25px] md:ml-1 sm:ml-2 xl:text-[17px]`}
                >
                    {weekdays[i]}
                </div>
            );
        }

        // Render empty days before the first day of the month
        for (let i = 0; i < firstDay; i++) {
            calendarDays.push(
                <div key={`empty-${i}`} className={`${styles.emptyDay}`}></div>
            );
        }

        // Render day numbers
        for (let day = 1; day <= daysInMonth; day++) {
            const thisDate = new Date(currentYear, currentMonth, day);
            const isSelected = selectedRange.some(
                (date) => date.toDateString() === thisDate.toDateString()
            );

            calendarDays.push(
                <button
                    key={`day-${day}`}
                    className={`${styles.calendarDay} ${
                        isSelected ? styles.selectedRange : ""
                    } xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded`}
                    onClick={() => handleDateClick(day)}
                >
                    {day}
                </button>
            );
        }

        return calendarDays;
    };

    //update selected days when resetDays changes
    useEffect(() => {
        if (resetDays) {
            setSelectedDays(resetDays);
            setRangeStart(null);
            setSelectedRange([]);
        }
    }, [resetDays]);

    // Render the Calendar component
    return (
        <>
            <div className='flex flex-col items-center'>
                <div className='flex flex-col-reverse items-center justify-center xl:w-[333px] xl:h-[243px] sm:w-[222px] sm:h-[180px]'>
                    <div className='flex justify-center items-center gap-2'>
                        <button
                            className='flex items-center'
                            onClick={handlePreviousMonth}
                        >
                            <Image
                                className='xl:w-[50px] xl:h-[50px] sm:w-[35px] md:h-[35px]'
                                src='/icons/previous.png'
                                width={50}
                                height={50}
                                alt='Previous'
                            />
                        </button>
                        <div className='flex flex-col items-center gap-3'>
                            <div className='calendar sm:h-[140px] xl:h-[200px] grid grid-cols-7 gap-1'>
                                {renderCalendar()}
                            </div>
                            <h2 className='text-xl xl:w-[200px] sm:w-[165px] flex justify-center gap-4 font-bold sm:text-[14px] xl:text-[18px]'>
                                <Dropdown
                                    options={Array.from(
                                        { length: 12 },
                                        (_, i) => i
                                    ).map((month) =>
                                        new Date(
                                            currentDate.getFullYear(),
                                            month
                                        ).toLocaleString("default", {
                                            month: "long",
                                        })
                                    )}
                                    selectedOption={getMonthName(currentDate)}
                                    setSelectedOption={(monthName) => {
                                        const month = new Date(
                                            Date.parse(monthName + " 1, 2012")
                                        ).getMonth();
                                        setCurrentDate(
                                            new Date(
                                                currentDate.getFullYear(),
                                                month
                                            )
                                        );
                                    }}
                                />
                                <Dropdown
                                    options={Array.from(
                                        { length: 6 },
                                        (_, i) =>
                                            i +
                                            Math.min(
                                                currentDate.getFullYear(),
                                                new Date().getFullYear()
                                            )
                                    )}
                                    selectedOption={currentDate
                                        .getFullYear()
                                        .toString()}
                                    setSelectedOption={(year) => {
                                        setCurrentDate(
                                            new Date(
                                                year,
                                                currentDate.getMonth()
                                            )
                                        );
                                    }}
                                />
                            </h2>
                        </div>
                        <button
                            className='flex items-center'
                            onClick={handleNextMonth}
                        >
                            <Image
                                className='xl:w-[50px] xl:h-[50px] sm:w-[35px] md:h-[35px]'
                                src='/icons/next.png'
                                width={50}
                                height={50}
                                alt='Next'
                            />
                        </button>
                    </div>
                </div>
            </div>
        </>
    );
};

export default Calendar;
