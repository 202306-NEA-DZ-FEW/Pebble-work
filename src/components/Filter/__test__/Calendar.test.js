import Calendar from "../Calendar";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";

describe("Calendar Component", () => {
    const mockProps = {
        checkEvents: jest.fn(),
        resetDays: {},
    };

    const renderComponent = (props = {}) => {
        render(<Calendar {...mockProps} {...props} />);
    };

    test("should render component", () => {
        render(<Calendar />);
        expect(screen.getByAltText("Previous")).toBeInTheDocument();
        expect(screen.getByAltText("Next")).toBeInTheDocument();
        expect(renderComponent()).toMatchSnapshot();
    });

    test("should handles date click", async () => {
        renderComponent();
        const dayElement = screen.getByRole("link", { name: "1" });
        fireEvent.click(dayElement);

        expect(mockProps.checkEvents).toHaveBeenCalledWith(expect.any(String));
    });
});
