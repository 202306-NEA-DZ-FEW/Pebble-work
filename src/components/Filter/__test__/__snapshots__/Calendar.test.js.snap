// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Component should render component 1`] = `undefined`;
